File: masv101.py
# ######################################################################################
#
# masv101.py
#
# Version: MASV v.1.0.1
#
# Description:
#   This script performs denoising of sequencing data from a FASTA file.
#   It identifies unique sequences, dereplicates them, and then classifies them
#   as either valid sequence variants or noise based on k-mer profile comparisons
#   and abundance ratios.
#
# Usage:
#   python masv101.py -i <input.fasta> -f <degree_of_freedom> -a <abundance_ratio> -s <save_spurious_as_variant>
#
# Arguments:
#   -i, --input_fasta: Path to the input FASTA file.
#   -f, --freedom: An integer representing the degree of freedom for noise filtering.
#   -a, --abundance_ratio: A float representing the minimum abundance ratio (parent/noise) required for filtering.
#   -s, --save_spurious: A boolean flag (True/False) to save sequences classified as 'SPURIOUS VARIANT' to the variants output file.
#
# ######################################################################################

import argparse, sys, time
from collections import Counter

# --- Argument Parsing ---
# Sets up the command-line interface for the user to provide input files and parameters.
parser = argparse.ArgumentParser(description="MASV v.1.0.1: A high-resolution denoising script for amplicon data.")
parser.add_argument('-i', dest='input_fasta', type=str, required=True,
                    help='Path to the input FASTA file.')
parser.add_argument('-f', dest='freedom', type=int, required=False, default=1,
                    help="Degree of freedom (integer). A multiplier for the denoising thresholds. "
                         "'-f 1' is the most stringent (default). '-f 2' is more lenient, etc.")
parser.add_argument('-a', dest='abundance_ratio', type=float, required=False, default=1.45,
                    help="Minimum abundance ratio (Parent Sequence Size / Noise Sequence Size) "
                         "required to classify a sequence as noise. Default is 1.45.")
parser.add_argument('-s', dest='save_spurious', type=lambda x: (x.lower() == 'true'), required=False, default=False,
                    help="Boolean flag (True/False). If True, singleton sequences classified as 'SPURIOUS VARIANT' "
                         "are saved to the 'variants.fa' file. If False, they are saved to 'noise.fa'. Default is False.")
args = parser.parse_args()


class Sequence:
    """
    Represents a biological sequence and its associated k-mer profile.

    This class stores the sequence information, calculates its k-mer profile,
    and provides a method to compare its profile with another sequence.
    The sequence is "doubled" (e.g., 'A' becomes 'AA') before k-mer generation
    to handle short sequences and minimize edge effects.
    """

    def __init__(self, title, sequence, k=2):
        """
        Initializes a Sequence object.

        Args:
            title (str): The FASTA header/title for the sequence.
                         The size (abundance) is parsed from this title.
            sequence (str): The DNA sequence string.
            k (int): The size of the k-mers to generate.
        """
        self.title = title  # Sequence identifier from FASTA header
        self.size = int(title.rsplit("=", 1)[-1][:-1]) # Abundance, parsed from title
        self.sequence = sequence.upper()  # The DNA sequence, converted to uppercase
        self.length = len(sequence)  # Length of the original sequence
        self.k = k  # k-mer size
        # Create a "doubled" sequence to enhance k-mer profile differences
        self.doubled_sequence = self.sequence.replace('A', 'AA').replace('C', 'CC').replace('T', 'TT').replace('G', 'GG')
        # Generate the k-mer profile from the doubled sequence
        self.kmers = self.generate_kmers(self.doubled_sequence, k)

    def generate_kmers(self, sequence, k):
        """
        Generates a k-mer profile from a sequence.

        Args:
            sequence (str): The sequence to process.
            k (int): The size of the k-mers.

        Returns:
            dict: A dictionary mapping each k-mer to its frequency count.
                  Only k-mers with valid DNA characters (A, C, T, G) are included.
        """
        kmers = {}
        valid_chars = {"A", "C", "T", "G"}
        for i in range(len(sequence) - k + 1):
            kmer = sequence[i:i + k]
            # Ensure all characters in the k-mer are valid DNA bases
            if all(c in valid_chars for c in kmer):
                kmers[kmer] = kmers.get(kmer, 0) + 1
        return kmers

    def compare_kmer_profiles(self, other):
        """
        Compares the k-mer profile of this sequence with another sequence.

        This method calculates the absolute difference in counts for each k-mer
        between the two profiles. It distinguishes between "perfect" k-mers
        (homopolymers like 'AA', 'CC') and "imperfect" k-mers.

        Args:
            other (Sequence): The other Sequence object to compare against.

        Returns:
            list: A list containing three integers:
                  [total_mismatch_count, perfect_kmer_mismatch, imperfect_kmer_mismatch]
        """
        mismatch_count_all = 0
        mismatch_count_actg = 0

        # Define perfect k-mers (homopolymers of length 2)
        perfect_kmers = {'AA', 'CC', 'TT', 'GG'}

        # Get the union of all k-mers present in either sequence
        keys_union = set(self.kmers) | set(other.kmers)

        for kmer in keys_union:
            # Calculate the absolute difference in frequency for the current k-mer
            diff = abs(self.kmers.get(kmer, 0) - other.kmers.get(kmer, 0))
            mismatch_count_all += diff
            # Check if the k-mer is one of the "perfect" k-mers
            if kmer in perfect_kmers:
                mismatch_count_actg += diff

        # Calculate the mismatch count for "imperfect" k-mers
        mismatch_count_other = mismatch_count_all - mismatch_count_actg

        return [mismatch_count_all, mismatch_count_actg, mismatch_count_other]


def dereplicate_fasta(fasta_file):
    """
    Reads a FASTA file, finds unique sequences, and counts their occurrences.

    Args:
        fasta_file (str): The path to the input FASTA file.

    Returns:
        dict: A dictionary mapping a new unique identifier (e.g., "Uniq0001;size=100;")
              to each unique sequence.
    """
    global total_input_seqs

    seq_counter = Counter()
    seq_chunks = []
    # Read the FASTA file line by line
    with open(fasta_file, 'r') as file:
        for line in file:
            line = line.strip()
            if line.startswith('>'):
                # When a new header is found, process the previous sequence
                if seq_chunks:
                    seq = ''.join(seq_chunks)
                    seq_counter[seq] += 1
                    seq_chunks = []
            else:
                # Append sequence lines to the chunk
                seq_chunks.append(line)
        # Process the last sequence in the file
        if seq_chunks:
            seq = ''.join(seq_chunks)
            seq_counter[seq] += 1

    # Sort sequences by abundance in descending order
    sorted_seqs = seq_counter.most_common()
    # Create new headers for unique sequences including their size (count)
    uniq_sequences = {f"Uniq{i + 1:04d};size={count};": seq
                      for i, (seq, count) in enumerate(sorted_seqs)}

    total_input_seqs = sum(seq_counter.values())
    print(f"Total input sequences: {total_input_seqs}")
    print(f"Dereplicated sequences: {len(uniq_sequences)}")

    return uniq_sequences


def sort_variants(seq1, seq2):
    """
    Compares two sequences to determine if one is a noisy variant of the other.

    A sequence (seq2) is classified as noise if it has a similar length,
    is significantly less abundant than the other sequence (seq1), and their
    k-mer profiles are very similar based on predefined thresholds.

    Args:
        seq1 (Sequence): The more abundant sequence (potential parent).
        seq2 (Sequence): The less abundant sequence (potential noisy variant).
    """
    global variants, noise, filtered, sequences, count, fx, ax

    # Condition 1: Length difference must be at most 1 base.
    if abs(seq1.length - seq2.length) <= 1:
        # Condition 2: Abundance ratio must be significant (parent must be >= ax more abundant).
        # Use the global abundance ratio 'ax' (from -a argument)
        if seq1.size / seq2.size >= ax:
            mismatch_count = seq1.compare_kmer_profiles(seq2)
            # Precompute mismatch values for clarity
            p = mismatch_count[1]   # Perfect k-mer mismatch
            im = mismatch_count[2]  # Imperfect k-mer mismatch
            imp = im - p            # Difference between imperfect and perfect mismatches
            len_diff = abs(seq1.length - seq2.length)
            # Condition 3: K-mer profile differences must be within thresholds defined by freedom factor `fx`.
            if p <= 4 * fx and im <= 4 * fx and imp <= 2 * fx:
                # If all conditions are met, classify seq2 as noise
                noise[seq2.title] = seq2.sequence
                filtered[seq2.title] = [seq2.title, seq1.title, "NOISY VARIANT", p, im, len_diff]
                count = count + 1


def write_results(variants, noise, filtered):
    """
    Writes the classification results to output files.

    - asv_tab.txt: A table summarizing the classification of each sequence.
    - variants.fa: A FASTA file of sequences classified as valid variants.
    - noise.fa: A FASTA file of sequences classified as noise.

    Args:
        variants (dict): Dictionary of valid sequence variants.
        noise (dict): Dictionary of noisy sequences.
        filtered (dict): Dictionary containing classification details for the table.
    """
    global weights
    # Write the summary table (ASV table)
    asv_table = 'asv_tab.txt'
    with open(asv_table, 'w') as at:
        at.write('title\tclosest neighbor(s)\tdescription\tperfect k-mer\timperfect k-mer\tlength difference\n')
        for k, v in filtered.items():
            at.write(
                str(v[0]) + '\t' + str(v[1]) + '\t' + str(v[2]) + '\t' + str(v[3]) + '\t' + str(v[4]) + '\t' + str(v[5]) + '\n')

    # Write the FASTA file for valid variants
    out_fasta_variants = 'variants.fa'
    with open(out_fasta_variants, 'w') as of:
        for i in range(len(variants)):
            title = str(list(variants.keys())[i])
            of.write('>' + title + 'neighbors=' + str(weights.get(title, 0)) + ';' + '\n')
            of.write(str(list(variants.values())[i]) + '\n')

    # Write the FASTA file for noisy sequences
    out_fasta_noise = 'noise.fa'
    with open(out_fasta_noise, 'w') as of:
        for i in range(len(noise)):
            of.write('>' + str(list(noise.keys())[i]) + '\n')
            of.write(str(list(noise.values())[i]) + '\n')


def progress_bar(i):
    """
    Displays a simple progress bar in the console.

    Args:
        i (int): The current iteration index.
    """
    progress = (i + 1) / len(sequences)
    bar_length = 40  # Visual length of the progress bar
    filled_length = int(bar_length * progress)
    bar = '#' * filled_length + '-' * (bar_length - filled_length)
    percent = progress * 100
    # Write the progress bar to the same line in the console
    sys.stdout.write(f"\r[{bar}] {percent:.1f}%")
    sys.stdout.flush()


# --- Main Execution ---
if __name__ == "__main__":
    print('\nscript MASV v.1.0.1')
    start_time = time.time()

    # Get arguments from command line
    input_fasta = args.input_fasta
    fx = int(args.freedom)
    # Get the new abundance ratio argument
    ax = args.abundance_ratio
    # Get the new spurious variant saving flag
    save_spurious = args.save_spurious

    # Step 1: Dereplicate the input FASTA file
    uniq_fasta = dereplicate_fasta(input_fasta)

    # Step 2: Create Sequence objects for each unique sequence
    sequences = [Sequence(title, seq) for title, seq in uniq_fasta.items()]

    # Initialize dictionaries to store results
    filtered = {} # Stores detailed classification info for the output table
    variants = {} # Stores sequences identified as true variants
    noise = {}    # Stores sequences identified as noise
    weights = {}  # Stores the count of noisy neighbors for each variant

    print('Identifying sequence variants:')
    # Step 3: Compare all unique sequences against each other
    for i, seq1 in enumerate(sequences):
        count = 0
        # Check if seq1 has not already been classified as noise
        if seq1.title not in filtered.keys():
            # Compare seq1 with all subsequent sequences in the list
            for j, seq2 in enumerate(sequences):
                # Ensure we don't compare a sequence to itself and that seq2 is not already classified
                if i < j and seq2.title not in filtered.keys():
                    sort_variants(seq1, seq2)

            # Store the number of noisy sequences found for seq1
            weights[seq1.title] = count

            # Classify seq1 as a variant if its abundance is 2 or more
            if seq1.size >= 2:
                variants[seq1.title] = seq1.sequence
                filtered[seq1.title] = [seq1.title, '*', "VARIANT", '*', '*', '*']
            # Sequences with size < 2 are not considered variants at this stage
            else:
                pass
        # Update the progress bar after processing each sequence
        progress_bar(i)

    # Step 4: Final classification for any remaining sequences
    # Any sequence not classified as a variant or noise is considered a "spurious variant"
    for seq1 in sequences:
        if seq1.title not in filtered.keys():
            # Use the 'save_spurious' flag (-s argument) to decide where to save the sequence
            if save_spurious:
                # Save to variants output
                variants[seq1.title] = seq1.sequence
            else:
                # Save to noise output (default behavior)
                noise[seq1.title] = seq1.sequence
            # The classification in the table remains the same
            filtered[seq1.title] = [seq1.title, '*', "SPURIOUS VARIANT", '*', '*', '*']

    # Step 5: Write all results to output files
    write_results(variants, noise, filtered)

    # --- Script Completion ---
    end_time = time.time()
    total_time = end_time - start_time
    print(f"\nDetected sequence variants: {len(variants)}")
    print(f"Script execution time: {total_time:.2f} seconds ({total_time / 60:.2f} minutes)\n")

File: masv101.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <unordered_map>
#include <map>
#include <cmath>
#include <algorithm>
#include <thread>
#include <mutex>
#include <atomic>
#include <chrono>
#include <memory>

// --- Global Structures & Config ---
const int PERFECT_IDX[] = {0, 5, 10, 15};
const int IMPERFECT_IDX[] = {1, 2, 3, 4, 6, 7, 8, 9, 11, 12, 13, 14};

inline int base_to_val(char c) {
    switch (c) {
        case 'A': return 0;
        case 'C': return 1;
        case 'G': return 2;
        case 'T': return 3;
        default: return -1;
    }
}

inline int kmer_to_idx(char b1, char b2) {
    int v1 = base_to_val(b1);
    int v2 = base_to_val(b2);
    if (v1 == -1 || v2 == -1) return -1;
    return (v1 << 2) | v2;
}

struct SequenceData {
    std::string title;
    std::string seq;
    int size;
    int length;
    std::vector<int> kmers;
};

// --- Helper Functions ---
std::vector<int> generate_kmer_vector(const std::string& seq, const std::string& title) {
    std::vector<int> vector(16, 0);
    for (char c : seq) {
        int val = base_to_val(c);
        if (val == -1) {
            std::cerr << "\nError processing " << title << ": Ambiguous nucleotide detected.\n";
            exit(1);
        }
        vector[(val << 2) | val]++; 
    }
    for (size_t i = 0; i < seq.length() - 1; ++i) {
        int idx = kmer_to_idx(seq[i], seq[i + 1]);
        if (idx != -1) vector[idx]++;
    }
    return vector;
}

std::vector<SequenceData> dereplicate_fasta(const std::string& filename) {
    std::unordered_map<std::string, int> seq_counter;
    std::vector<std::string> order;
    
    std::ifstream file(filename);
    if (!file.is_open()) {
        std::cerr << "Error: Could not open file " << filename << "\n";
        exit(1);
    }

    std::string line, current_seq = "";
    int total_count = 0;

    while (std::getline(file, line)) {
        if (line.empty()) continue;
        if (line[0] == '>') {
            if (!current_seq.empty()) {
                if (seq_counter.find(current_seq) == seq_counter.end()) order.push_back(current_seq);
                seq_counter[current_seq]++;
                total_count++;
                current_seq.clear();
            }
        } else {
            for (char &c : line) c = toupper(c);
            current_seq += line;
        }
    }
    if (!current_seq.empty()) {
        if (seq_counter.find(current_seq) == seq_counter.end()) order.push_back(current_seq);
        seq_counter[current_seq]++;
        total_count++;
    }
    file.close();

    std::vector<std::pair<std::string, int>> sorted_seqs;
    sorted_seqs.reserve(order.size());
    for (const auto& seq : order) {
        sorted_seqs.push_back({seq, seq_counter[seq]});
    }
    std::sort(sorted_seqs.begin(), sorted_seqs.end(), [](const auto& a, const auto& b) {
        return a.second > b.second;
    });

    std::vector<SequenceData> data;
    data.reserve(sorted_seqs.size());
    for (size_t i = 0; i < sorted_seqs.size(); ++i) {
        char title_buf[64];
        sprintf(title_buf, "Uniq%04zu;size=%d;", i + 1, sorted_seqs[i].second);
        
        SequenceData sd;
        sd.title = std::string(title_buf);
        sd.seq = sorted_seqs[i].first;
        sd.size = sorted_seqs[i].second;
        sd.length = static_cast<int>(sd.seq.length());
        data.push_back(std::move(sd));
    }

    std::cout << "Total input sequences: " << total_count << "\n";
    std::cout << "Dereplicated sequences: " << data.size() << "\n";
    return data;
}

// Global outputs protected by a single Mutex (Only used when a match is actually found)
std::mutex output_mutex;
std::unordered_map<std::string, std::vector<std::string>> filtered_map;
std::unordered_map<std::string, std::string> variants_map;
std::unordered_map<std::string, std::string> noise_map;
std::unordered_map<std::string, int> weights_map;

struct NoiseMeta {
    int target_idx;
    int p_sum;
    int im_sum;
    int len_diff;
};

int main(int argc, char* argv[]) {
    std::string input_fasta = "";
    int fx = 1;
    float ax = 1.45f;
    bool save_spurious = false;
    int num_threads = 2;

    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        if (arg == "-i" && i + 1 < argc) input_fasta = argv[++i];
        else if (arg == "-f" && i + 1 < argc) fx = std::stoi(argv[++i]);
        else if (arg == "-a" && i + 1 < argc) ax = std::stof(argv[++i]);
        else if (arg == "-s" && i + 1 < argc) save_spurious = (std::string(argv[++i]) == "True");
        else if (arg == "-t" && i + 1 < argc) num_threads = std::stoi(argv[++i]);
    }

    if (input_fasta.empty()) {
        std::cerr << "Usage: " << argv[0] << " -i <input.fasta> [-f freedom] [-a abundance] [-s True/False] [-t threads]\n";
        return 1;
    }

    std::cout << "\nMASV v.1.0.1 (Multi-Threaded C++ Engine)\n";
    auto start_time = std::chrono::high_resolution_clock::now();

    auto dataset = dereplicate_fasta(input_fasta);
    size_t N = dataset.size();
    if (N == 0) return 0;

    std::cout << "Generating K-mer feature profiles across " << num_threads << " thread(s)...\n";
    std::vector<std::thread> kmer_threads;
    std::atomic<size_t> kmer_index(0);

    for (int t = 0; t < num_threads; ++t) {
        kmer_threads.emplace_back([&]() {
            while (true) {
                size_t curr = kmer_index.fetch_add(1, std::memory_order_relaxed);
                if (curr >= N) break;
                dataset[curr].kmers = generate_kmer_vector(dataset[curr].seq, dataset[curr].title);
            }
        });
    }
    for (auto& th : kmer_threads) th.join();

    std::cout << "Building algorithmic lookup buckets...\n";
    std::unordered_map<int, std::vector<int>> length_buckets;
    for (int i = 0; i < static_cast<int>(N); ++i) {
        length_buckets[dataset[i].length].push_back(i);
    }

    std::cout << "Identifying sequence variants...\n";
    
    // THE FIX: Lock-Free Atomic Array
    std::unique_ptr<std::atomic<bool>[]> active(new std::atomic<bool>[N]);
    for(size_t i = 0; i < N; ++i) {
        active[i].store(true, std::memory_order_relaxed);
    }

    std::atomic<size_t> main_index(0);
    std::atomic<size_t> progress_counter(0);
    std::vector<std::thread> worker_threads;

    for (int t = 0; t < num_threads; ++t) {
        worker_threads.emplace_back([&]() {
            
            // THE FIX: Pre-allocate memory ONCE per thread, outside the loop
            std::vector<int> candidates;
            candidates.reserve(10000); 
            std::vector<size_t> flagged_noise_indices;
            flagged_noise_indices.reserve(10000);
            std::vector<NoiseMeta> noise_metadata;
            noise_metadata.reserve(10000);

            while (true) {
                size_t i = main_index.fetch_add(1, std::memory_order_relaxed);
                if (i >= N) break;

                // Ultra-fast lock-free read
                if (!active[i].load(std::memory_order_relaxed)) {
                    progress_counter.fetch_add(1, std::memory_order_relaxed);
                    continue;
                }

                candidates.clear();
                flagged_noise_indices.clear();
                noise_metadata.clear();

                int L_i = dataset[i].length;
                for (int target_L = L_i - 1; target_L <= L_i + 1; ++target_L) {
                    auto it = length_buckets.find(target_L);
                    if (it != length_buckets.end()) {
                        for (int idx : it->second) {
                            if (idx > static_cast<int>(i)) candidates.push_back(idx);
                        }
                    }
                }

                int loop_count = 0;

                for (int target_idx : candidates) {
                    // Ultra-fast lock-free read
                    if (!active[target_idx].load(std::memory_order_relaxed)) continue;

                    if ((static_cast<float>(dataset[i].size) / dataset[target_idx].size) < ax) continue;

                    int p_sum = 0;
                    int im_sum = 0;
                    bool failed = false;

                    for (int k = 0; k < 16; ++k) {
                        int diff = std::abs(dataset[i].kmers[k] - dataset[target_idx].kmers[k]);
                        if (k == 0 || k == 5 || k == 10 || k == 15) {
                            p_sum += diff;
                            if (p_sum > 4 * fx) { failed = true; break; } 
                        } else {
                            im_sum += diff;
                            if (im_sum > 4 * fx) { failed = true; break; } 
                        }
                    }
                    if (failed || (im_sum - p_sum) > 2 * fx) continue;

                    flagged_noise_indices.push_back(target_idx);
                    int len_diff = std::abs(dataset[i].length - dataset[target_idx].length);
                    noise_metadata.push_back({target_idx, p_sum, im_sum, len_diff});
                    loop_count++;
                }

                // Ultra-fast lock-free write (Deactivate noise sequences immediately)
                for (size_t idx : flagged_noise_indices) {
                    active[idx].store(false, std::memory_order_relaxed);
                }

                // Only lock when writing confirmed variants to the global dictionary
                if (!noise_metadata.empty() || dataset[i].size >= 2) {
                    std::lock_guard<std::mutex> lock(output_mutex);
                    weights_map[dataset[i].title] = loop_count;

                    for (const auto& meta : noise_metadata) {
                        std::string t_title = dataset[meta.target_idx].title;
                        noise_map[t_title] = dataset[meta.target_idx].seq;
                        filtered_map[t_title] = {t_title, dataset[i].title, "NOISY VARIANT", std::to_string(meta.p_sum), std::to_string(meta.im_sum), std::to_string(meta.len_diff)};
                    }

                    if (dataset[i].size >= 2) {
                        variants_map[dataset[i].title] = dataset[i].seq;
                        filtered_map[dataset[i].title] = {dataset[i].title, "*", "VARIANT", "*", "*", "*"};
                    }
                }

                size_t done = progress_counter.fetch_add(1, std::memory_order_relaxed) + 1;
                if (done % std::max((size_t)1, N / 100) == 0 || done == N) {
                    std::cout << "\rProcessing updates: " << (done * 100 / N) << "% Completed." << std::flush;
                }
            }
        });
    }
    for (auto& th : worker_threads) th.join();

    for (size_t i = 0; i < N; ++i) {
        if (filtered_map.find(dataset[i].title) == filtered_map.end()) {
            if (save_spurious) {
                variants_map[dataset[i].title] = dataset[i].seq;
            } else {
                noise_map[dataset[i].title] = dataset[i].seq;
            }
            filtered_map[dataset[i].title] = {dataset[i].title, "*", "SPURIOUS VARIANT", "*", "*", "*"};
        }
    }

    std::cout << "\nWriting data outputs to files...\n";
    std::ofstream tab_file("asv_tab.txt");
    tab_file << "title\tclosest neighbor(s)\tdescription\tperfect k-mer\timperfect k-mer\tlength difference\n";
    for (size_t i = 0; i < N; ++i) {
        auto it = filtered_map.find(dataset[i].title);
        if (it != filtered_map.end()) {
            tab_file << it->second[0] << "\t" << it->second[1] << "\t" << it->second[2] << "\t" 
                     << it->second[3] << "\t" << it->second[4] << "\t" << it->second[5] << "\n";
        }
    }
    tab_file.close();

    std::ofstream var_file("variants.fa");
    for (size_t i = 0; i < N; ++i) {
        auto it = variants_map.find(dataset[i].title);
        if (it != variants_map.end()) {
            var_file << ">" << it->first << "neighbors=" << weights_map[it->first] << ";\n" << it->second << "\n";
        }
    }
    var_file.close();

    std::ofstream noise_file("noise.fa");
    for (size_t i = 0; i < N; ++i) {
        auto it = noise_map.find(dataset[i].title);
        if (it != noise_map.end()) {
            noise_file << ">" << it->first << "\n" << it->second << "\n";
        }
    }
    noise_file.close();

    auto end_time = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> total_time = end_time - start_time;
    
    std::cout << "\nDetected sequence variants: " << variants_map.size() << "\n";
    std::cout << "Script execution time: " << total_time.count() << " seconds.\n\n";

    return 0;
}